src/V3DfgCache.h

```diff
@@ -0,0 +1,402 @@
+                 std::forward_as_tuple(nullptr))
+        .first->second;
+}
+
+inline DfgVertexUnary*& getEntry(CacheUnary& cache, AstNodeDType*, DfgVertex* src0p) {
+    return cache
+        .emplace(std::piecewise_construct,  //
+                 std::forward_as_tuple(src0p),  //
+                 std::forward_as_tuple(nullptr))
+        .first->second;
+}
+
+inline DfgVertexBinary*& getEntry(CacheBinary& cache, AstNodeDType*, DfgVertex* src0p,
+                                  DfgVertex* src1p) {
+    return cache
+        .emplace(std::piecewise_construct,  //
+                 std::forward_as_tuple(src0p, src1p),  //
+                 std::forward_as_tuple(nullptr))
+        .first->second;
+}
+
+inline DfgVertexTernary*& getEntry(CacheTernary& cache, AstNodeDType*, DfgVertex* src0p,
+                                   DfgVertex* src1p, DfgVertex* src2p) {
+    return cache
+        .emplace(std::piecewise_construct,  //
+                 std::forward_as_tuple(src0p, src1p, src2p),  //
+                 std::forward_as_tuple(nullptr))
+        .first->second;
+}
+
+// These return a reference to the mapped entry, inserting a nullptr if not yet exists
+inline CacheSel::iterator find(CacheSel& cache, AstNodeDType* dtypep, DfgVertex* src0p,
+                               uint32_t lsb) {
+    UASSERT_OBJ(VN_IS(dtypep, BasicDType), dtypep, "non-packed has no 'width()'");
+    const uint32_t width = dtypep->width();
+    return cache.find({src0p, lsb, width});
+}
+
+inline CacheUnary::iterator find(CacheUnary& cache, AstNodeDType*, DfgVertex* src0p) {
+    return cache.find({src0p});
+}
+
+inline CacheBinary::iterator find(CacheBinary& cache, AstNodeDType*, DfgVertex* src0p,
+                                  DfgVertex* src1p) {
+    return cache.find({src0p, src1p});
+}
+
+inline CacheTernary::iterator find(CacheTernary& cache, AstNodeDType*, DfgVertex* src0p,
+                                   DfgVertex* src1p, DfgVertex* src2p) {
+    return cache.find({src0p, src1p, src2p});
+}
+
+// These set the operands of a new vertex
+inline void setOperands(DfgSel* vtxp, DfgVertex* fromp, uint32_t lsb) {
+    vtxp->fromp(fromp);
+    vtxp->lsb(lsb);
+}
+
+inline void setOperands(DfgVertexUnary* vtxp, DfgVertex* src0p) {  //
+    vtxp->relinkSource<0>(src0p);
+}
+
+inline void setOperands(DfgVertexBinary* vtxp, DfgVertex* src0p, DfgVertex* src1p) {
+    vtxp->relinkSource<0>(src0p);
+    vtxp->relinkSource<1>(src1p);
+}
+
+inline void setOperands(DfgVertexTernary* vtxp, DfgVertex* src0p, DfgVertex* src1p,
+                        DfgVertex* src2p) {
+    vtxp->relinkSource<0>(src0p);
+    vtxp->relinkSource<1>(src1p);
+    vtxp->relinkSource<2>(src2p);
+}
+
+// Get or create (and insert) vertex with given operands
+template <typename Vertex, typename Cache, typename... Operands>
+inline Vertex* getOrCreate(DfgGraph& dfg, FileLine* flp, AstNodeDType* dtypep, Cache& cache,
+                           Operands... operands) {
```

Comment:
Are you sure ... used here is C++14?

Reply:
Fairly. It would be quite simpler with C++17 `if constexpr`